From ef93807e200b47e15390c8a38045e93749a3ee3e Mon Sep 17 00:00:00 2001
From: Alex Gough <ajgo@chromium.org>
Date: Wed, 6 Apr 2022 17:31:24 -0700
Subject: [PATCH 1/3] Prevent duplicate attributes in mojo

Before:

[Honk=Foo,Honk=Bar]
...

was accepted (probably only one of the values would be used).

Now:

```
Traceback (most recent call last):
  File "../../mojo/public/tools/mojom/mojom_parser.py", line 500, in <module>
    Run(sys.argv[1:])
...
    raise Exception("Duplicate key (%s) in attribute list" % attribute.key)
Exception: Duplicate key (Honk) in attribute list
```

Bug: 931315
Change-Id: I794af3ad9edc468631deed7d6d7a762915305fea
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3575275
Reviewed-by: Oksana Zhuravlova <oksamyt@chromium.org>
Commit-Queue: Alex Gough <ajgo@chromium.org>
Cr-Commit-Position: refs/heads/main@{#989706}


CrOS-Libchrome-Original-Commit: 2af71f7ced8596f34124a30883385d855c5a62f8
---
 .../tools/mojom/mojom/generate/translate.py      | 10 +++++++---
 .../mojom/mojom/generate/translate_unittest.py   | 16 ++++++++++++++++
 2 files changed, 23 insertions(+), 3 deletions(-)

diff --git a/mojo/public/tools/mojom/mojom/generate/translate.py b/mojo/public/tools/mojom/mojom/generate/translate.py
index 92233c95d9..3e06751da4 100644
--- a/mojo/public/tools/mojom/mojom/generate/translate.py
+++ b/mojo/public/tools/mojom/mojom/generate/translate.py
@@ -139,9 +139,13 @@ def _AttributeListToDict(attribute_list):
   if attribute_list is None:
     return None
   assert isinstance(attribute_list, ast.AttributeList)
-  # TODO(vtl): Check for duplicate keys here.
-  return dict(
-      [(attribute.key, attribute.value) for attribute in attribute_list])
+  attributes = dict()
+  for attribute in attribute_list:
+    if attribute.key in attributes:
+      raise Exception("Duplicate key (%s) in attribute list" % attribute.key)
+    else:
+      attributes[attribute.key] = attribute.value
+  return attributes
 
 
 builtin_values = frozenset([
diff --git a/mojo/public/tools/mojom/mojom/generate/translate_unittest.py b/mojo/public/tools/mojom/mojom/generate/translate_unittest.py
index e4cd9e6075..cd50cc8ab4 100644
--- a/mojo/public/tools/mojom/mojom/generate/translate_unittest.py
+++ b/mojo/public/tools/mojom/mojom/generate/translate_unittest.py
@@ -88,3 +88,19 @@ class TranslateTest(unittest.TestCase):
     ])
     with self.assertRaises(Exception):
       translate.OrderedModule(tree, "mojom_tree", [])
+
+  def testDuplicateAttributesException(self):
+    tree = ast.Mojom(None, ast.ImportList(), [
+        ast.Union(
+            "FakeUnion",
+            ast.AttributeList([
+                ast.Attribute("key1", "value"),
+                ast.Attribute("key1", "value")
+            ]),
+            ast.UnionBody([
+                ast.UnionField("a", None, None, "int32"),
+                ast.UnionField("b", None, None, "string")
+            ]))
+    ])
+    with self.assertRaises(Exception):
+      translate.OrderedModule(tree, "mojom_tree", [])
-- 
2.31.0

